lib/actions/posts.ts

```diff
@@ -7,107 +7,233 @@ import { connectToDB } from "~/utils/database";
+async function _getPostsByPage(
+  page: number,
+  id: string,
+  limit: number = postsPerPage
+) {
+  try {
+    await connectToDB();
 
-  //get the posts of the user from the database
-  const posts = await Note.find({ creator: id })
-    .skip((page - 1) * limit)
-    .limit(limit);
+    //get the posts of the user from the database
+    const posts = await Note.find({ creator: id })
+      .sort({ createdAt: -1 })
+      .skip((page - 1) * limit)
+      .limit(limit);
 
-  return posts;
+    return posts;
+  } catch (err) {
+    console.log(err);
+    return null;
+  }
+}
+
+async function _getAllContributions(id: string) {
+  try {
+    await connectToDB();
+
+    //get the posts of the user from the database
+    const posts = await Note.find({ creator: id });
+
+    let contributionsMap: { [key: string]: number } = {};
+
+    posts.forEach((post) => {
+      const createdAt = format(post.createdAt, "yyyy-MM-dd");
+      const updatedAt = format(post.updatedAt, "yyyy-MM-dd");
+      const dates = [createdAt, updatedAt];
+      for (let date of dates) {
+        if (contributionsMap[date]) {
+          contributionsMap[date] += 1;
+        } else {
+          contributionsMap[date] = 1;
+        }
+      }
+    });
+
+    const contributions: Contribution[] = Object.keys(contributionsMap).map(
+      (date, index) => ({
+        id: index + 1,
+        date: date,
+        count: contributionsMap[date],
+      })
+    );
+
+    return { userID: id, contributions: contributions };
+  } catch (err) {
+    console.log(err);
+    return null;
+  }
 }
 
 async function _getAllOtherPosts(postIDs: string[], creatorID: string) {
+  try {
+    await connectToDB();
+
+    //get the posts of the user from the database
+    const posts = await Note.find({
+      creator: creatorID,
+      _id: { $nin: postIDs },
+    });
+    const postTitlesAndIDs = posts.map((post) => {
+      return { title: post.title, id: post.id } as {
+        title: string;
+        id: string;
+      };
+    });
+
+    return postTitlesAndIDs as { title: string; id: string }[];
+  } catch (err) {
+    console.log(err);
+    return null;
+  }
+}
+
+async function _searchPosts(searchTerm: string) {
   await connectToDB();
 
-  //get the posts of the user from the database
-  const posts = await Note.find({
-    creator: creatorID,
-    _id: { $nin: postIDs },
-  });
-  const postTitlesAndIDs = posts.map((post) => {
-    return { title: post.title, id: post.id } as { title: string; id: string };
-  });
+  if (searchTerm.length < 3) {
+    return [];
+  }
 
-  return postTitlesAndIDs as { title: string; id: string }[];
+  try {
+    const posts = await Note.find({
+      $or: [
+        { title: { $regex: searchTerm, $options: "i" } },
+        { description: { $regex: searchTerm, $options: "i" } },
+        { content: { $regex: searchTerm, $options: "i" } },
+      ],
+    }).sort({ createdAt: -1 });
+
+    console.log("Posts found: ", posts.length);
+    return posts;
```

Comment:
```suggestion
   return JSON.parse(JSON.stringify(posts));
```

Reply:
Seems to solve my intermittent "maximum callstack" errors. Anytime we're sending data from a server action to a client component or server component to a client component we need to clean up the object that comes back from MongoDB.

We won't have to worry about that with other ORMS! This is particularly a MongoDB/Next issue